src/zkevm_specs/opcode/byte.py

```diff
@@ -0,0 +1,33 @@
+from typing import Sequence
+
+from ..encoding import U8, is_circuit_code, u8s_to_u256, u256_to_u8s
+
+@is_circuit_code
+def check_byte(
+    v8s: Sequence[U8],
+    i8s: Sequence[U8],
+    r8s: Sequence[U8],
+):
+    assert len(v8s) == len(i8s) == len(r8s) == 32
+
+    # Any index value >= 256 always returns all zeros
+    msb_sum_zero = sum(i8s[1:]) == 0
+
+    # Byte 0:
+    r = r8s[0]
+    for i in range(32):
+        selected = i8s[0] == i
+        r = r - (v8s[i] * msb_sum_zero * selected)
```

Comment:
Why we are having `msb_sum_zero` here? Is this a trick to avoid creating another polynomial to check `msb_sum_zero`?

Reply:
This is done here because only the LSB value of the index is used to check if the byte should be copied. So the LSB value of the index may be [0..32] which would mean we have to copy the byte value, but of course if any of the other index byte values are non-zero the index is actually a much bigger value so we should never copy any of the bytes. So `msb_sum_zero` just checks if any of those byte values are zero or non-zero once, and the we use it here as an extra condition for the byte copy.

Reply:
I see in your circuit PR https://github.com/appliedzkp/zkevm-circuits/pull/106 you apply IsZero gadget to `msb_sum_zero` to make it 0 or 1.
But you need an extra column for IsZero gadget of msb_sum_zero`.

What do you think we do the following? Does it achieve the same result?

```python
# 1 poly (no new column for `msb_sum_zero`)
assert sum(i8s[1:]) == 0
...

    r = r - (v8s[i] * selected)  # (no `msb_sum_zero` here)
```

Reply:
Ah sorry, I see the problem with my proposed solution.
So in EVM, the non-zero MSB in `index` is still valid and will not raise any exception. It only causes the copied `result` to be zero.
We should add this note to the spec.